File: Lander/core/src/com/nopalsoft/lander/game/TiledMapManagerBox2d.java
package com.nopalsoft.lander.game;

import com.badlogic.gdx.maps.Map;
import com.badlogic.gdx.maps.MapLayer;
import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.MapObjects;
import com.badlogic.gdx.maps.MapProperties;
import com.badlogic.gdx.maps.objects.CircleMapObject;
import com.badlogic.gdx.maps.objects.PolygonMapObject;
import com.badlogic.gdx.maps.objects.PolylineMapObject;
import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.maps.objects.TextureMapObject;
import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType;
import com.badlogic.gdx.physics.box2d.ChainShape;
import com.badlogic.gdx.physics.box2d.CircleShape;
import com.badlogic.gdx.physics.box2d.FixtureDef;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.Shape;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.utils.Logger;
import com.nopalsoft.lander.game.objetos.Bomba;
import com.nopalsoft.lander.game.objetos.Estrella;
import com.nopalsoft.lander.game.objetos.Gas;
import com.nopalsoft.lander.game.objetos.Laser;
import com.nopalsoft.lander.game.objetos.Nave;
import com.nopalsoft.lander.game.objetos.Plataforma;

public class TiledMapManagerBox2d {

    private final WorldGame oWorld;
    private final World oWorldBox;
    private final float m_units;
    private final Logger logger;
    private final FixtureDef defaultFixture;

    public TiledMapManagerBox2d(WorldGame oWorld, float unitsPerPixel) {
        this.oWorld = oWorld;
        oWorldBox = oWorld.oWorldBox;
        m_units = unitsPerPixel;
        logger = new Logger("MapBodyManager", 1);

        defaultFixture = new FixtureDef();
        defaultFixture.density = 1.0f;
        defaultFixture.friction = .5f;
        defaultFixture.restitution = 0.0f;
    }

    public void createObjetosDesdeTiled(Map map) {
        crearFisicos(map);
        crearInteaccion(map);
    }

    private void crearInteaccion(Map map) {
        MapLayer layer = map.getLayers().get("interaccion");

        if (layer == null) {
            logger.error("layer " + "interaccion" + " no existe");
            return;
        }

        MapObjects objects = layer.getObjects();
        for (MapObject object : objects) {
            if (object instanceof TextureMapObject) {
                continue;
            }

            MapProperties properties = object.getProperties();
            String tipo = (String) properties.get("type");
            switch (tipo) {
                case "estrella":
                    crearEstrella(object);
                    break;
                case "gas":
                    crearGas(object);
                    break;
                case "laser":
                    crearLaser(object);
                    break;
                case "bomba":
                    crearBomba(object);
                    break;
            }
        }
    }

    private void crearFisicos(Map map) {
        MapLayer layer = map.getLayers().get("fisicos");

        if (layer == null) {
            logger.error("layer " + "fisicos" + " no existe");
            return;
        }

        MapObjects objects = layer.getObjects();

        for (MapObject object : objects) {
            if (object instanceof TextureMapObject) {
                continue;
            }

            MapProperties properties = object.getProperties();
            String tipo = (String) properties.get("type");
            if (tipo == null)
                continue;
            else if (tipo.equals("inicio") || tipo.equals("fin")) {
                creatPlataformas(object, tipo);
            }

            Shape shape;
            if (object instanceof RectangleMapObject) {
                shape = getRectangle((RectangleMapObject) object);
            } else if (object instanceof PolygonMapObject) {
                shape = getPolygon((PolygonMapObject) object);
            } else if (object instanceof PolylineMapObject) {
                shape = getPolyline((PolylineMapObject) object);
            } else if (object instanceof CircleMapObject) {
                shape = getCircle((CircleMapObject) object);
            } else {
                logger.error("non suported shape " + object);
                continue;
            }

            defaultFixture.shape = shape;

            BodyDef bodyDef = new BodyDef();
            bodyDef.type = BodyType.StaticBody;

            Body body = oWorldBox.createBody(bodyDef);
            body.createFixture(defaultFixture);

            body.setUserData(tipo);

            defaultFixture.shape = null;
            defaultFixture.isSensor = false;
            defaultFixture.filter.groupIndex = 0;
            shape.dispose();
        }
    }

    private void creatPlataformas(MapObject object, String tipo) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        PolygonShape polygon = new PolygonShape();
        Vector2 size = new Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units);
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f);
        defaultFixture.shape = polygon;

        BodyDef bodyDef = new BodyDef();
        bodyDef.type = BodyDef.BodyType.StaticBody;
        Body body = oWorldBox.createBody(bodyDef);
        body.createFixture(defaultFixture);

        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        float height = (rectangle.height * m_units * 0.5f);
        float width = (rectangle.width * m_units * 0.5f);

        Plataforma plataforma = new Plataforma(x, y, width, height);

        body.setUserData(plataforma);
        oWorld.arrPlataformas.add(plataforma);

        if (tipo.equals("inicio"))
            crearNave(plataforma);
        else if (tipo.equals("fin"))
            plataforma.isFinal = true;
    }

    private void crearNave(Plataforma plataforma) {
        Nave oNave = new Nave(plataforma.position.x, plataforma.position.y + plataforma.size.y / 2);

        BodyDef bd = new BodyDef();
        bd.position.x = oNave.position.x;
        bd.position.y = oNave.position.y;
        bd.type = BodyType.DynamicBody;
        Body oBody = oWorldBox.createBody(bd);

        PolygonShape shape = new PolygonShape();
        shape.setAsBox(Nave.WIDTH / 2f, Nave.HEIGHT / 2f);// Lo puse mejor con vertices pa que no tuviera esquinas picudas y tratar de minimizar

        FixtureDef fixture = new FixtureDef();
        fixture.shape = shape;
        fixture.density = Nave.DENSIDAD_INICIAL;
        fixture.restitution = 0;
        fixture.friction = .5f;
        oBody.createFixture(fixture);

        oBody.setUserData(oNave);
        oBody.setBullet(true);

        shape.dispose();

        oWorld.oNave = oNave;
    }

    private void crearEstrella(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;
        float height = (rectangle.height * m_units * 0.5f);
        float width = (rectangle.width * m_units * 0.5f);

        Estrella obj = new Estrella(x, y, width, height);
        BodyDef bd = new BodyDef();
        bd.position.y = obj.position.y;
        bd.position.x = obj.position.x;
        bd.type = BodyType.StaticBody;

        PolygonShape pies = new PolygonShape();
        pies.setAsBox(.1f, .1f);

        FixtureDef fixture = new FixtureDef();
        fixture.shape = pies;
        fixture.density = .5f;
        fixture.restitution = 0f;
        fixture.friction = 0f;
        fixture.isSensor = true;
        // fixture.filter.groupIndex = CONTACT_CORREDORES;

        Body oBody = oWorldBox.createBody(bd);
        oBody.createFixture(fixture);

        oBody.setUserData(obj);

        oWorld.arrEstrellas.add(obj);
        pies.dispose();
    }

    private void crearGas(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;
        float height = (rectangle.height * m_units * 0.5f);
        float width = (rectangle.width * m_units * 0.5f);

        Gas obj = new Gas(x, y, width, height);
        BodyDef bd = new BodyDef();
        bd.position.y = obj.position.y;
        bd.position.x = obj.position.x;
        bd.type = BodyType.StaticBody;

        PolygonShape pies = new PolygonShape();
        pies.setAsBox(.1f, .1f);

        FixtureDef fixture = new FixtureDef();
        fixture.shape = pies;
        fixture.density = .5f;
        fixture.restitution = 0f;
        fixture.friction = 0f;
        fixture.isSensor = true;
        // fixture.filter.groupIndex = CONTACT_CORREDORES;

        Body oBody = oWorldBox.createBody(bd);
        oBody.createFixture(fixture);

        oBody.setUserData(obj);

        oWorld.arrGas.add(obj);
        pies.dispose();
    }

    private void crearLaser(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        String direccion = object.getProperties().get("direccion").toString();
        float timeOFF, timeOffActual;

        timeOFF = Float.parseFloat(object.getProperties().get("tiempoApagado").toString());
        timeOffActual = Float.parseFloat(object.getProperties().get("tiempoApagadoActual").toString());


        Laser obj = new Laser(x, y, rectangle.getWidth() * m_units, rectangle.height * m_units, timeOFF, timeOffActual, direccion);
        BodyDef bd = new BodyDef();
        bd.position.y = obj.position.y;
        bd.position.x = obj.position.x;
        bd.type = BodyType.StaticBody;

        PolygonShape pies = new PolygonShape();
        pies.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units);

        FixtureDef fixture = new FixtureDef();
        fixture.shape = pies;
        fixture.density = .5f;
        fixture.restitution = 0f;
        fixture.friction = 0f;
        fixture.isSensor = true;
        // fixture.filter.groupIndex = CONTACT_CORREDORES;

        Body oBody = oWorldBox.createBody(bd);
        oBody.createFixture(fixture);

        oBody.setUserData(obj);

        oWorld.arrLaser.add(obj);
        pies.dispose();
    }

    private void crearBomba(MapObject object) {
        Rectangle rectangle = ((RectangleMapObject) object).getRectangle();
        float x = (rectangle.x + rectangle.width * 0.5f) * m_units;
        float y = (rectangle.y + rectangle.height * 0.5f) * m_units;

        Bomba obj = new Bomba(x, y, object.getProperties().get("direccion").toString());

        BodyDef bd = new BodyDef();
        bd.position.y = obj.position.y;
        bd.position.x = obj.position.x;
        bd.type = BodyType.DynamicBody;

        CircleShape pies = new CircleShape();
        pies.setRadius(.2f);
        // pies.setAsBox(.05f, .05f);

        FixtureDef fixture = new FixtureDef();
        fixture.shape = pies;
        fixture.density = 15f;
        fixture.restitution = 0f;
        fixture.friction = 0f;
        // fixture.filter.groupIndex = CONTACT_CORREDORES;

        Body oBody = oWorldBox.createBody(bd);
        oBody.setGravityScale(0);
        oBody.createFixture(fixture);

        oBody.setUserData(obj);
        oWorld.arrBombas.add(obj);

        pies.dispose();
    }

    private Shape getRectangle(RectangleMapObject rectangleObject) {
        Rectangle rectangle = rectangleObject.getRectangle();
        PolygonShape polygon = new PolygonShape();
        Vector2 size = new Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units);
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f);
        return polygon;
    }

    private Shape getCircle(CircleMapObject circleObject) {
        Circle circle = circleObject.getCircle();
        CircleShape circleShape = new CircleShape();
        circleShape.setRadius(circle.radius * m_units);
        circleShape.setPosition(new Vector2(circle.x * m_units, circle.y * m_units));
        return circleShape;
    }

    private Shape getPolygon(PolygonMapObject polygonObject) {
        PolygonShape polygon = new PolygonShape();
        float[] vertices = polygonObject.getPolygon().getVertices();
        float[] worldVertices = new float[vertices.length];
        float yLost = polygonObject.getPolygon().getY() * m_units;
        float xLost = polygonObject.getPolygon().getX() * m_units;

        for (int i = 0; i < vertices.length; ++i) {
            if (i % 2 == 0)
                worldVertices[i] = vertices[i] * m_units + xLost;
            else
                worldVertices[i] = vertices[i] * m_units + yLost;
        }
        polygon.set(worldVertices);

        return polygon;
    }

    private Shape getPolyline(PolylineMapObject polylineObject) {
        float[] vertices = polylineObject.getPolyline().getVertices();

        Vector2[] worldVertices = new Vector2[vertices.length / 2];
        float yLost = polylineObject.getPolyline().getY() * m_units;
        float xLost = polylineObject.getPolyline().getX() * m_units;
        for (int i = 0; i < vertices.length / 2; ++i) {
            worldVertices[i] = new Vector2();
            worldVertices[i].x = vertices[i * 2] * m_units + xLost;
            worldVertices[i].y = vertices[i * 2 + 1] * m_units + yLost;
        }
        ChainShape chain = new ChainShape();
        chain.createChain(worldVertices);
        return chain;
    }
}

File: Lander/core/src/com/nopalsoft/lander/game/TiledMapManagerBox2d.kt
package com.nopalsoft.lander.game

import com.badlogic.gdx.maps.Map
import com.badlogic.gdx.maps.MapObject
import com.badlogic.gdx.maps.objects.CircleMapObject
import com.badlogic.gdx.maps.objects.PolygonMapObject
import com.badlogic.gdx.maps.objects.PolylineMapObject
import com.badlogic.gdx.maps.objects.RectangleMapObject
import com.badlogic.gdx.maps.objects.TextureMapObject
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType
import com.badlogic.gdx.physics.box2d.ChainShape
import com.badlogic.gdx.physics.box2d.CircleShape
import com.badlogic.gdx.physics.box2d.FixtureDef
import com.badlogic.gdx.physics.box2d.PolygonShape
import com.badlogic.gdx.physics.box2d.Shape
import com.badlogic.gdx.utils.Logger
import com.nopalsoft.lander.game.objetos.Bomba
import com.nopalsoft.lander.game.objetos.Estrella
import com.nopalsoft.lander.game.objetos.Gas
import com.nopalsoft.lander.game.objetos.Laser
import com.nopalsoft.lander.game.objetos.Nave
import com.nopalsoft.lander.game.objetos.Plataforma

class TiledMapManagerBox2d(private val oWorld: WorldGame, private val m_units: Float) {
    private val oWorldBox = oWorld.oWorldBox
    private val logger = Logger("MapBodyManager", 1)
    private val defaultFixture = FixtureDef()

    init {
        defaultFixture.density = 1.0f
        defaultFixture.friction = .5f
        defaultFixture.restitution = 0.0f
    }

    fun createObjetosDesdeTiled(map: Map) {
        crearFisicos(map)
        crearInteaccion(map)
    }

    private fun crearInteaccion(map: Map) {
        val layer = map.layers.get("interaccion")

        if (layer == null) {
            logger.error("layer " + "interaccion" + " no existe")
            return
        }

        val mapLayerObjects = layer.objects
        for (layerObject in mapLayerObjects) {
            if (layerObject is TextureMapObject) {
                continue
            }

            val properties = layerObject.properties
            val tipo = properties.get("type") as String
            when (tipo) {
                "estrella" -> createStar(layerObject)
                "gas" -> createGas(layerObject)
                "laser" -> createLaser(layerObject)
                "bomba" -> createBomb(layerObject)
            }
        }
    }

    private fun crearFisicos(map: Map) {
        val layer = map.layers.get("fisicos")

        if (layer == null) {
            logger.error("layer " + "fisicos" + " no existe")
            return
        }

        val mapLayerObjects = layer.objects

        for (layerObject in mapLayerObjects) {
            if (layerObject is TextureMapObject) {
                continue
            }

            val properties = layerObject.properties
            val tipo = properties.get("type") as String?
            if (tipo == null) continue
            else if (tipo == "inicio" || tipo == "fin") {
                creatPlataformas(layerObject, tipo)
            }
            val shape = if (layerObject is RectangleMapObject) {
                getRectangle(layerObject)
            } else if (layerObject is PolygonMapObject) {
                getPolygon(layerObject)
            } else if (layerObject is PolylineMapObject) {
                getPolyline(layerObject)
            } else if (layerObject is CircleMapObject) {
                getCircle(layerObject)
            } else {
                logger.error("non suported shape $layerObject")
                continue
            }

            defaultFixture.shape = shape

            val bodyDef = BodyDef()
            bodyDef.type = BodyType.StaticBody

            val body = oWorldBox.createBody(bodyDef)
            body.createFixture(defaultFixture)

            body.userData = tipo

            defaultFixture.shape = null
            defaultFixture.isSensor = false
            defaultFixture.filter.groupIndex = 0
            shape.dispose()
        }
    }

    private fun creatPlataformas(platformMapObject: MapObject, tipo: String) {
        val rectangle = (platformMapObject as RectangleMapObject).rectangle
        val polygon = PolygonShape()
        val size = Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units)
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f)
        defaultFixture.shape = polygon

        val bodyDef = BodyDef()
        bodyDef.type = BodyType.StaticBody
        val body = oWorldBox.createBody(bodyDef)
        body.createFixture(defaultFixture)

        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        val height = (rectangle.height * m_units * 0.5f)
        val width = (rectangle.width * m_units * 0.5f)

        val plataforma = Plataforma(x, y, width, height)

        body.userData = plataforma
        oWorld.arrPlataformas.add(plataforma)

        if (tipo == "inicio") crearNave(plataforma)
        else if (tipo == "fin") plataforma.isFinal = true
    }

    private fun crearNave(plataforma: Plataforma) {
        val oNave = Nave(plataforma.position.x, plataforma.position.y + plataforma.size.y / 2)

        val bd = BodyDef()
        bd.position.x = oNave.position.x
        bd.position.y = oNave.position.y
        bd.type = BodyType.DynamicBody
        val oBody = oWorldBox.createBody(bd)

        val shape = PolygonShape()
        shape.setAsBox(Nave.WIDTH / 2f, Nave.HEIGHT / 2f) // Lo puse mejor con vertices pa que no tuviera esquinas picudas y tratar de minimizar

        val fixture = FixtureDef()
        fixture.shape = shape
        fixture.density = Nave.DENSIDAD_INICIAL
        fixture.restitution = 0f
        fixture.friction = .5f
        oBody.createFixture(fixture)

        oBody.userData = oNave
        oBody.isBullet = true

        shape.dispose()

        oWorld.oNave = oNave
    }

    private fun createStar(starMapObject: MapObject) {
        val rectangle = (starMapObject as RectangleMapObject).rectangle
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        val obj = Estrella(x, y)
        val bd = BodyDef()
        bd.position.y = obj.position.y
        bd.position.x = obj.position.x
        bd.type = BodyType.StaticBody

        val pies = PolygonShape()
        pies.setAsBox(.1f, .1f)

        val fixture = FixtureDef()
        fixture.shape = pies
        fixture.density = .5f
        fixture.restitution = 0f
        fixture.friction = 0f
        fixture.isSensor = true

        val oBody = oWorldBox.createBody(bd)
        oBody.createFixture(fixture)

        oBody.userData = obj

        oWorld.arrEstrellas.add(obj)
        pies.dispose()
    }

    private fun createGas(gasMapObject: MapObject) {
        val rectangle = (gasMapObject as RectangleMapObject).rectangle
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        val obj = Gas(x, y)
        val bd = BodyDef()
        bd.position.y = obj.position.y
        bd.position.x = obj.position.x
        bd.type = BodyType.StaticBody

        val pies = PolygonShape()
        pies.setAsBox(.1f, .1f)

        val fixture = FixtureDef()
        fixture.shape = pies
        fixture.density = .5f
        fixture.restitution = 0f
        fixture.friction = 0f
        fixture.isSensor = true

        val oBody = oWorldBox.createBody(bd)
        oBody.createFixture(fixture)

        oBody.userData = obj

        oWorld.arrGas.add(obj)
        pies.dispose()
    }

    private fun createLaser(laserMapObject: MapObject) {
        val rectangle = (laserMapObject as RectangleMapObject).rectangle
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        val direccion = laserMapObject.properties.get("direccion").toString()

        val timeOFF: Float = laserMapObject.properties.get("tiempoApagado").toString().toFloat()
        val timeOffActual: Float = laserMapObject.properties.get("tiempoApagadoActual").toString().toFloat()


        val obj = Laser(x, y, rectangle.getWidth() * m_units, rectangle.height * m_units, timeOFF, timeOffActual, direccion)
        val bd = BodyDef()
        bd.position.y = obj.position.y
        bd.position.x = obj.position.x
        bd.type = BodyType.StaticBody

        val pies = PolygonShape()
        pies.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units)

        val fixture = FixtureDef()
        fixture.shape = pies
        fixture.density = .5f
        fixture.restitution = 0f
        fixture.friction = 0f
        fixture.isSensor = true

        val oBody = oWorldBox.createBody(bd)
        oBody.createFixture(fixture)

        oBody.userData = obj

        oWorld.arrLaser.add(obj)
        pies.dispose()
    }

    private fun createBomb(bombMapObject: MapObject) {
        val rectangle = (bombMapObject as RectangleMapObject).rectangle
        val x = (rectangle.x + rectangle.width * 0.5f) * m_units
        val y = (rectangle.y + rectangle.height * 0.5f) * m_units

        val obj = Bomba(x, y, bombMapObject.properties.get("direccion").toString())

        val bd = BodyDef()
        bd.position.y = obj.position.y
        bd.position.x = obj.position.x
        bd.type = BodyType.DynamicBody

        val pies = CircleShape()
        pies.radius = .2f
        val fixture = FixtureDef()
        fixture.shape = pies
        fixture.density = 15f
        fixture.restitution = 0f
        fixture.friction = 0f

        val oBody = oWorldBox.createBody(bd)
        oBody.gravityScale = 0f
        oBody.createFixture(fixture)

        oBody.userData = obj
        oWorld.arrBombas.add(obj)

        pies.dispose()
    }

    private fun getRectangle(rectangleObject: RectangleMapObject): Shape {
        val rectangle = rectangleObject.rectangle
        val polygon = PolygonShape()
        val size = Vector2((rectangle.x + rectangle.width * 0.5f) * m_units, (rectangle.y + rectangle.height * 0.5f) * m_units)
        polygon.setAsBox(rectangle.getWidth() * 0.5f * m_units, rectangle.height * 0.5f * m_units, size, 0.0f)
        return polygon
    }

    private fun getCircle(circleObject: CircleMapObject): Shape {
        val circle = circleObject.circle
        val circleShape = CircleShape()
        circleShape.radius = circle.radius * m_units
        circleShape.position = Vector2(circle.x * m_units, circle.y * m_units)
        return circleShape
    }

    private fun getPolygon(polygonObject: PolygonMapObject): Shape {
        val polygon = PolygonShape()
        val vertices = polygonObject.polygon.vertices
        val worldVertices = FloatArray(vertices.size)
        val yLost = polygonObject.polygon.y * m_units
        val xLost = polygonObject.polygon.x * m_units

        for (i in vertices.indices) {
            if (i % 2 == 0) worldVertices[i] = vertices[i] * m_units + xLost
            else worldVertices[i] = vertices[i] * m_units + yLost
        }
        polygon.set(worldVertices)

        return polygon
    }

    private fun getPolyline(polylineObject: PolylineMapObject): Shape {
        val vertices = polylineObject.polyline.vertices

        val worldVertices = arrayOfNulls<Vector2>(vertices.size / 2)
        val yLost = polylineObject.polyline.y * m_units
        val xLost = polylineObject.polyline.x * m_units
        for (i in 0..<vertices.size / 2) {
            worldVertices[i] = Vector2()
            worldVertices[i]!!.x = vertices[i * 2] * m_units + xLost
            worldVertices[i]!!.y = vertices[i * 2 + 1] * m_units + yLost
        }
        val chain = ChainShape()
        chain.createChain(worldVertices)
        return chain
    }
}
